feat: CLREWriter + Proper "modded.res" + Patched space station

wackfx opened the pull request:
Hey there @WizzardMaker 👋 

First off, so cool that you're maintaining this! It solved so many issues I had on Windows 11 (freeze on music change, crashes ...).

In order to get my feet wet to the project, I decided to improve french translation.

You'll find:
- A CRLEWriter class that convert the .txt files into proper xxx_ger.res
- Used it to create a new modded_ger.res file to avoid any lawsuit with the official assets
- Started to convert the code to use this new file
- Added a new command `/updatedata` to automatically build .txt file to .res before running the game. It's the opposite of `/savedata`

You can see this in actions in mods/misc/modded_res with the options you added.

Here is the translations in action
![image](https://github.com/WizzardMaker/AirlineTycoon/assets/135170502/d6fe97b0-f65e-4da0-85b3-a01c1b8b20da)

> ⚠️ This would require to add the `/misc/modded_ger.res` to releases - but I believe it pave the way to more friendly modding and solves [issues like #50](https://github.com/WizzardMaker/AirlineTycoon/issues/50#issuecomment-1748501310)

I didn't convert ALL translations - as I didn't know if you were interested in this or not. Let me know, I will gladly patch all the added text so far into this new system.

Cya!

PS: Small heads up - this PR is on top of your fix/various branch

wackfx committed:
feat: CLREWriter
feat: Modded texts example (in a proper file)

Commit:
fix: comment

wackfx committed:
Update TeakLibW.h

Commit:
feat: Method to find a string in resource files
feat: Added a modded override check
feat: Patched space station prices + texts

wackfx commented:
I've added some stuff:
- Added a `Texte::FindS` that returns nullptr. This is useful to check whether a resource file has a translation
- Used this new `FindS` in `ChangeMoney` to check for an override in the new `modded_ger.res`. Fallback to StandardTexte like before
- This allowed for a neat trick to actually implement Space Station items in history 😏
- This actually patches #44 (found and patched this price bug as well)

![image](https://github.com/WizzardMaker/AirlineTycoon/assets/135170502/d9e948b4-2c36-4b15-aece-9f6abf25258d)

wackfx commented on mods/misc/modded_ger.res.txt:
```diff
@@ -0,0 +1,177 @@
+  D::# Sound:
+
+>Mony
+>>0008000
```
Those are starship overrides to have a history perfectly related to what you bought.

wackfx commented on src/Dialog.cpp:
```diff
@@ -5935,7 +5935,7 @@ BOOL CStdRaum::PreLButtonDown(CPoint point) {
                 } else {
                     (dynamic_cast<CNasa *>((qPlayer.DialogWin) != nullptr ? qPlayer.DialogWin : this))->KommVarTippNow = 12;
-                    qPlayer.AddRocketPart(Flag, RocketPrices[Index]);
+                    qPlayer.AddSpaceStationPart(Flag, id, StationPrices[Index]);
```
There was the #44 price bug. Added a new method to handle space station parts properly

wackfx commented on src/Global.cpp:
```diff
@@ -132,9 +132,10 @@ BOOL gRoomJustLeft; // TRUE, wenn Raum gerade verlassen wurde (wegen messagePump
-TEXTRES ETexte;        // Die Einheiten
+TEXTRES DialogTexte;    // Die Text-Ressourcen der Dialoge
+TEXTRES StandardTexte;  // Allgemeine Texte
+TEXTRES ModdedTexte;    // Modded Texts
```
Added a global modded texte that points to the new modded_ger.res

wackfx commented on src/Player.cpp:
```diff
@@ -223,8 +223,12 @@ void PLAYER::ChangeMoney(__int64 Money, SLONG Reason, const CString &Par1, char
+    char const *text = ModdedTexte.FindS(TOKEN_MONEY, Reason);
+    if (text == nullptr) {
+        text = StandardTexte.GetS(TOKEN_MONEY, Reason);
+    }
```
This performs the modded override. If we find a Money entry in the modded resource file, we use it. If not, we just fallback to the standard texte.

wackfx commented on src/TeakLibW.h:
```diff
@@ -729,9 +752,12 @@ class TEXTRES {
     char *GetP(ULONG, ULONG);
+    char *FindP(ULONG, ULONG);
     char *GetS(ULONG, ULONG);
+    char *FindS(ULONG, ULONG);
```
new methods that allows to search in texte files.
GetS would throw if not found - so i've added a FindS that will just return nullptr if not found.

This allows us to implement proper modded resource override.

wackfx commented on src/TeakLibW/TeakFile.cpp:
```diff
@@ -162,11 +164,11 @@ void CRLEReader::SaveAsPlainText() {
 
         // TEAKFILE file(fn, TEAKFILE_WRITE);
-        FILE *fp = fopen(fn, "w");
+        FILE *fp = fopen(fn, "wb");
```
Added the `b` flag to write as binary.

`fputc` would add a `\r\n` on `\n` - Resulting in line break multiplication (`\r\r\n`) when read / wrote mutiple times.

With the binary flag - it doesn't do that. File is still perfectly readable by any text editor.

wackfx commented on src/TeakLibW/TeakFile.cpp:
```diff
@@ -257,10 +259,126 @@ BUFFER_V<BYTE> LoadCompleteFile(char const *path) {
 }
+
+
+CRLEWriter::CRLEWriter(const char *path) : Ctx(nullptr), Version(0x102), Key(0xA5), Magic("xtRLE"), Path(path), Sequence() {}
```
Actual `.txt` to `.res` shenanigans.

You can use both flags `/updatedata /savedata` to test it - it works really well.

Commit:
Reworked LoadCompleteFile to avoid open FD issues
Integrated override system into TeakLib
Cleaned up code

Commit:
overrides files

wackfx commented:
Actually converted (and translated) all texts from code into this new system.

Commit:
[fix] Patched graphical issue on route conf.
[feat] Added an option to confire RoutePrice increment (asked on Discord)
[feat] Added some shorter helper (1000000 > 1.000k) (see Shorten1000erDots)
[fix] Graphical issues with big numbers on BUY_KEROSENE screen
[fix] German translations + corrected some translations

wackfx commented:
new stuff
[fix] Patched graphical issue on route conf - some longer text would mix with price. Now patched
![image](https://github.com/WizzardMaker/AirlineTycoon/assets/135170502/6098cf5d-c0e8-4ae4-9aa6-90e7a8189b3a)


[feat] Added an option to confire RoutePrice increment (asked on Discord)

![image](https://github.com/WizzardMaker/AirlineTycoon/assets/135170502/bcc7e641-9021-4781-a09a-a8d5cb5a799f)


[fix] Graphical issues with big numbers on BUY_KEROSENE screen - Now avoiding text to be scrambled with big numbers

![image](https://github.com/WizzardMaker/AirlineTycoon/assets/135170502/c83390cb-cf56-42ef-9c92-5a165ce70ad3)

Commit:
feat: Plane availability (broker / museum or both)

wackfx commented:
feat: Plane availability feature #45
Allows to add a new "availability" column to planetyp.csv in order to make the planes available on MUSEUM, BROKER or both.
By default, or if the column doesn't exists (backcompatibility), the plane will be available in both. 
![image](https://github.com/WizzardMaker/AirlineTycoon/assets/135170502/c87d543e-6c1d-4a7f-b2b0-a1a96502677c)

Commit:
Patched an issue on loading a network save file without session name

Commit:
[feat] Options to customize cities / day
[feat] Improved Board / Post It system to allow for more entries

WizzardMaker commented:
Happy to see so much progress!

I'll take a look at it in detail, as soon as I come back from my Christmas vacation, next year!

---

Your fixes to the text solution are very welcome!
Having the strings in code was always a thorn in my eyes

Regarding your proposal to include modded files as separate entities - I already played with the idea of doing something similar. The only hurdle I saw at the time was the complexity of including those in a release on GitHub. There were already "complications" with instructing people to include the binary dependencies.
For a similar project, I developed a companion application that downloaded/patched all necessary files and kept those up to date on a new version/release. The problem I see here though is the tech stack of this project. We are using C++, which is notorious for being difficult in working with HTTPS web requests across all the platforms, and GitHub only offers an HTTPS endpoint for their API.

But that's a story for a different PR - I think we can just bundle the additional files with the .exe, as they don't depend on any platform and are rather lightweight. Did you take a look at the "release" GitHub workflow, that bundles everything for releases? That would be the perfect place to zip all files up for release. I could also implement that, if you're not comfortable with workflows.

Any particular reason, why you chose to label those files as "mods" and not something like "patch". I see mods as more or less optional changes to a base, whereas patches are changes that shouldn't be optional - as are the changes you propose/added. Just a thought from my side, and nothing neck breaking for merging this.

----
The plan is to first merge my changes from "various-fixes" to develop, then check out your PR, merge that into develop and then merge develop finally into main.


Do you plan on adding anything else to this PR? It's getting rather large 😅

WizzardMaker commented:
Could you also document your new features in the `README.md` (don't forget to credit yourself) when you are done?

WizzardMaker commented:
I took a look at your new config entries - some of them are a bit troublesome for multiplayer, as these could lead to desync issues down the line on different settings. Especially because settings aren't synchronized across players.

This isn't an issue for now and probably only a few people are ever going to be changing these, but one should maybe add a system to check for a config mismatch. There should maybe be a split between gameplay/difficulty settings and client side settings, to allow for more granular/easy mismatch detection. But that's also an issue for later - if at all

wackfx commented:
Hey @WizzardMaker, thank you for your comments!

---- 

About the text solution:
- Agreed that "mods" is a bit too broad - I will rename it
- Indeed, having a proper automatic patching system would be really great. I can take a look into it.
- I will also work on adding them to the release.zip through the workflow - this shouldn't be an issue.

PR Size: You're very right, I might start creating new PR(s) on top of this one from now on to ease the review process 😅 

Readme: No worries, will document everything!

Config: You're totally right, I'm not used to MP on this game. I believe a first fix would be to 1) use default values when in MP and only use those values for SP and maybe down the line 2) find a way to synchronise those parameters from the host maybe ?

WizzardMaker commented:
> Indeed, having a proper automatic patching system would be really great. I can take a look into it.

A possible solution would be a "launcher"/"version checker" that updates the game and all files with it.

I already implemented such a solution in C# for the Settlers 4 HD patch - but that application would need to be rewritten for a cross plattform UI, as currently only windows is supported.
The underlying code however can be easily adapted (the base for that updater is: https://github.com/bezzad/Downloader) and I think it would suffice to maybe just have a command line version for now

WizzardMaker commented:
> Config: You're totally right, I'm not used to MP on this game. I believe a first fix would be to 1) use default values when in MP and only use those values for SP and maybe down the line 2) find a way to synchronise those parameters from the host maybe ?

I think it would be better to maybe add a notification to the multiplayer menu, when changes to those settings are detected. This makes multiplayer with those settings possible and notifies users that they should check their settings.

Synchronization is rather difficult due to the way how the settings class is implemented - that makes it rather cumbersome to change settings just for a single game.

Commit:
WIP: cmake + patched build

Commit:
CMake multiplatform build for patches + improvements

Commit:
cmakewip branch

Commit:
Cmake

Commit:
Cmake + new workflow

Commit:
Updated workflow

wackfx commented:
new stuff:
- Updated CMake to centralize cross-platform configuration ( with Windows support )
- Updated workflow to combine all into .zip / .tar that includes compiled patched .ger files - Ready to copy/paste
- Creating draft releases on push to main / develop. It will also build artifacts on pull requests to easily test them
![image](https://github.com/WizzardMaker/AirlineTycoon/assets/135170502/40ba13fe-51ce-4f8d-9a2e-b191a759f00b)
- You can see it in action [here](https://github.com/wackfx/AirlineTycoon/actions/runs/7364473059)
![image](https://github.com/WizzardMaker/AirlineTycoon/assets/135170502/b60e0d64-5bcd-425c-bc30-0c8abcf8f5c4)
- I believe having this enforces that we have linux building - and using the same toolchain to keep a certain consistency across platforms
- Added installation / building instructions for Visual Studio Code / CMake
- Added a `gamefile`  folder that will be used while dev. with CMake to automatically apply gameassets into target build
- Updated Visual Studio configuration to update patch files
- Reorganized README, BUILDING.md is now seperated as I believe most of the users wouldn't care + it centralises the building informations.

That would be the final missing piece for this (except with the automatic "update" system, which I believe would be better suited for another PR).
Next commits would only be small stuff, like the multiplayer notice when config was changed + README new additions

WizzardMaker commented:
Looks great!
Love the unified workflow file, very clean.

Two things I noticed after a quick glance:
1. The artifacts are archived twice, once in the pipeline and a second time by GitHub's download process
![image](https://github.com/WizzardMaker/AirlineTycoon/assets/7768485/62e0553a-7c09-41ba-a3ff-2e79eb07b706)

2. The artifacts contain the `AT.json` config file. This could lead to problems with inexperienced users that just copy/paste the archives contents into the game folder, thus overriding their configuration. The game should autogenerate with defaults by itself, when trying to access an undefined config.

wackfx commented:
> The artifacts are archived twice, once in the pipeline and a second time by GitHub's download process

Yeah, this unfortunately is a limitation brought by upload-artifact and download-artifact actions
- action/upload-artifact always [zip (or rezip) the artifacts given at the end](https://github.com/actions/upload-artifact?tab=readme-ov-file#zip-archives).
- actions/download-artifact actually uses the raw (given to upload; unzipped) files because files [aren't zipped yet](https://github.com/actions/download-artifact/issues/143)

Scenario 1 - we don't pack before upload-artifact
- We would upload a folder containing the files - which will then be packaged into a .zip by the action / workflow/
- This would result in a simple .zip (which is good) in the build artifacts
- However, actions/download would pick up the raw folders / files - not the zip in the release job
- We would have to deal with the zipping out of the build job (running on each os) - increasing overall complexity

Scenario 2 - we pack before
- We upload a built .zip / .tar with action/upload-artifact - resulting in this weird .zip.zip and .tar.zip in build artifacts (which is a con)
- We end up with the correct "raw upload" .zip / .tar files in the release job - used to publish the draft release. We don't have to deal with any more complexity there.

I decided to go with the solution 2, as:
- It adds some frictions only to us, developers, when we want to use the built artifacts
- This might be patched [by the PR I linked previously](https://github.com/actions/download-artifact/issues/143) - looks like they changed stuff in v4 and might be able to download the final .zip instead of what was uploaded "raw"
- It thus would be way easier and allow us to implement Solution 1 without the repack issues: action/upload would upload the raw files  in a folder (resulting in single .zip artifacts) and download would just use the resulting .zip files.

WDYT ?

> The artifacts contain the AT.json config file. This could lead to problems with inexperienced users that just copy/paste the archives contents into the game folder, thus overriding their configuration. The game should autogenerate with defaults by itself, when trying to access an undefined config.

You're very correct about this - it's an overlook. I will patch this to avoid having the AT.json in archives

WizzardMaker commented:
The important thing is, that the release files aren't zipped twice.
Everything else is a nice to have. So, do as you see fit

Commit:
fix: Avoid writing options everywhere
fix: Reorganized Takeoff.cpp to allow for one-shot CLI run
feat: Allow for CLI one-shot run without creating any AT.json artifact

Commit:
Added to readme

Commit:
Corrected workflow

Commit:
patched workflow

wackfx commented:
Ok, so, about the AT.json.
It was generated because I'm using the executable as a CLI to generate the patched .res files with a command `/update-patched-files-only`  that should transform .txt files to the final .res file and exit.

However, there was some class (like Sim) that was constructed (being global singleton) and that wrote AT.json, even if the game didn't start.

I've rewrote a bit Takeoff + Checkup and modified Sim to avoid various AT.json writing, we can thus use the CLI mechanism to actually execute part of C++ logic from our workflow, without creating artifacts files like this.

About the multiplayer notice for the settings: I can't make network game work on my computer anymore 😢 I tested with my build and your 1.7.1 Pre-Release (which did work previously when I tried) and no kudos. When I try to join with IP on Player 2 - it just doesn't do anything.

If you have any pointers about this issue, that'll be amazing. If not that's ok, I will try again later, maybe it's a random windows issue because my PC's been running for a couple of days now, will try a reboot tonight.

Commit:
More translations patchs

Commit:
Patched security office closed text (missing translation and bugging)
Patched "Abandonner la ligen" > "Abandonner la ligne"

WizzardMaker commented:
You need to rebase your branch on the new develop branch head

wackfx committed:
MacOS build

wackfx commented:
Hey @WizzardMaker!
I rebased my branch on top of the develop branch. No conflicts whatsoever as commits from your branch should be the same.

Also, I patched the MacOS build. I still have some issues with the MacOS json writer/reader but I believe this should be easy to fix

Commit:
patched i64d in MacOS AT.json

wackfx commented on src/Checkup.cpp:
```diff
@@ -91,7 +91,7 @@ void CRegistryAccess::Close() {
 //--------------------------------------------------------------------------------------------
 bool CRegistryAccess::IsOpen() { return (settingsJSON != nullptr); }
 
-void CRegistryAccess::WriteFile() { 
```
This is just some automatic reformatting

Commit:
build on push

Commit:
workflo

Commit:
workflow patch

Commit:
MacOS build + added patch files to build

Commit:
patch: workflows
